src/msb_radix.zig

```diff
@@ -0,0 +1,196 @@
+        return truncate(u8, (@bitCast(U, @as(T, std.math.minInt(T))) ^ @bitCast(U, x)) >> shift);
+    }
+    return truncate(u8, @bitCast(U, x) >> shift);
+}
+
+inline fn readTwoBytes(
+    comptime T: type,
+    comptime idx: usize,
+    x: T
+) u16 {
+    const U = comptime std.meta.Int(.unsigned, std.meta.bitCount(T));
+    const shift = comptime (@sizeOf(T) - 2 - idx) * 8;
+    if (idx == 0) {
+        return truncate(u16, (@bitCast(U, @as(T, std.math.minInt(T))) ^ @bitCast(U, x)) >> shift);
+    }
+    return truncate(u16, @bitCast(U, x) >> shift);
+}
+
+fn radixSortByBytesAdaptive(
+    comptime T: type,
+    comptime Ubucket: type,
+    comptime idx: usize,
+    comptime array_is_final_destination: bool,
+    comptime BYTES_PER_LEVEL: usize,
+    noalias array: [*]T,
+    noalias scratch: [*]T,
+    arr_len: usize,
+    noalias buckets: [*]Ubucket,
+) void {
+    comptime {
+        std.debug.assert(BYTES_PER_LEVEL == 1 or BYTES_PER_LEVEL == 2);
+    }
+
+    var i: usize = 0;
+    var lo: usize = 0;
+    const buckets_len = comptime if (BYTES_PER_LEVEL == 2) 0x10000 else 0x100;
+    const bucketsize = buckets;
+    const bucketindex = buckets + buckets_len;
+
+    const readBucket = comptime if (BYTES_PER_LEVEL == 2) readTwoBytes else readOneByte;
+
+    i = 0;
+    while (i < buckets_len) : (i += 1) {
+        bucketsize[i] = 0;
+    }
+
+    i = 0;
+    while (i < arr_len) : (i += 1) {
+        const bucket = readBucket(T, idx, array[i]);
+        bucketsize[bucket] += 1;
+    }
+
+    bucketindex[0] = 0;
+    i = 1;
+    while (i < buckets_len) : (i += 1) {
+        bucketindex[i] = bucketindex[i-1] + bucketsize[i-1];
+    }
+
+    i = 0;
+    while (i < arr_len) : (i += 1) {
+        const bucket = readBucket(T, idx, array[i]);
+        scratch[bucketindex[bucket]] = array[i];
+        bucketindex[bucket] += 1;
+    }
+
+    const next_idx = comptime idx + BYTES_PER_LEVEL;
+    if (next_idx >= @sizeOf(T)) {
+        if (array_is_final_destination) {
+            i = 0;
+            while (i < arr_len) : (i += 1) {
+                array[i] = scratch[i];
+            }
+        }
+        return;
+    }
+
+    i = 0;
+    while (i < buckets_len) : (i += 1) {
+        const len = bucketsize[i];
+        const hi = lo + len;
+        if (BYTES_PER_LEVEL == 2 and next_idx+1 < @sizeOf(T) and len >= 0x10000) {
+            radixSortByBytesAdaptive(
+                T, Ubucket, next_idx, !array_is_final_destination, 2,
+                scratch+lo, array+lo, len, bucketindex,
+            );
+        } else if (len > INSSORT_CUTOFF) {
+            radixSortByBytesAdaptive(
+                T, Ubucket, next_idx, !array_is_final_destination, 1,
+                scratch+lo, array+lo, len, bucketindex,
+            );
+        } else if (len > 1) {
+            if (array_is_final_destination) {
+                insSortIntoOtherArray(T, scratch+lo, array+lo, len);
+            } else {
+                insSort(T, scratch+lo, len);
```

Comment:
passing *many pointers* arrays here isn't perfection. i would like to use `src/insertion.zig` but i'll merge this and fix it later

Reply:
This code is based on some code that was ported to C++ and saw a measurable speedup from not using slices. For example, the main sorting routine would take 3 pointers and 3 usize's if you used slices, but only look at 1 of the usize's. For insSort it makes no difference though